packages/test/test-end-to-end-tests/src/test/dataMigration.spec.ts

```diff
@@ -0,0 +1,425 @@
+		dataObjectFactoryV1,
+		[[dataObjectFactoryV1.type, Promise.resolve(dataObjectFactoryV1)]],
+		undefined,
+		undefined,
+		runtimeOptions,
+	);
+
+	const runtimeFactoryV1ToV2 = new ContainerRuntimeFactoryWithDefaultDataStore(
+		dataObjectFactoryV1ToV2,
+		[[dataObjectFactoryV1ToV2.type, Promise.resolve(dataObjectFactoryV1ToV2)]],
+		undefined,
+		undefined,
+		runtimeOptions,
+	);
+
+	const runtimeFactoryV2 = new ContainerRuntimeFactoryWithDefaultDataStore(
+		dataObjectFactoryV2,
+		[[dataObjectFactoryV2.type, Promise.resolve(dataObjectFactoryV2)]],
+		undefined,
+		undefined,
+		runtimeOptions,
+	);
+
+	const deferred = new Deferred<{
+		summarizer: ContainerRuntime;
+		readonlyRuntime: ContainerRuntime;
+	}>();
+
+	const dataMigrationRuntimeFactoryV2 = new ContainerRuntimeFactoryWithDataMigration(
+		runtimeFactoryV1,
+		runtimeFactoryV1ToV2,
+		runtimeFactoryV2,
+		deferred,
+	);
+
+	const createV1Container = async (): Promise<IContainer> =>
+		provider.createContainer(runtimeFactoryV1, undefined, { package: "v1" });
+
+	const loadV2Container = async (summaryVersion: string): Promise<IContainer> =>
+		provider.loadContainer(
+			dataMigrationRuntimeFactoryV2,
+			{ options: { cache: false } },
+			{ headers: { [LoaderHeader.version]: summaryVersion, [LoaderHeader.cache]: false } },
+			[
+				[{ package: "v1" }, dataMigrationRuntimeFactoryV2],
+				[{ package: "v2" }, dataMigrationRuntimeFactoryV2],
+			],
+		);
+
+	const loadV2ContainerOnly = async (summaryVersion: string): Promise<IContainer> =>
+		provider.loadContainer(
+			runtimeFactoryV2,
+			{ options: { cache: false } },
+			{ headers: { [LoaderHeader.version]: summaryVersion, [LoaderHeader.cache]: false } },
+			[[{ package: "v1" }, runtimeFactoryV2]],
+		);
+
+	beforeEach(async () => {
+		provider = getTestObjectProvider({ syncSummarizer: true });
+	});
+
+	it("Can migrate with ContainerRuntimeFactory", async () => {
+		// Setup container with basic dataObject
+		const container = await createV1Container();
+		const startObject = await requestFluidObject<DataObjectV1>(container, "/");
+		startObject.rootDDS.set("an", "op");
+		const waitForSummary = new Deferred<any>();
+		startObject.containerRuntime.on("op", (op) => {
+			if (op.type === MessageType.SummaryAck) {
+				waitForSummary.resolve(op.contents);
+			}
+		});
+		const firstSummaryAck = await waitForSummary.promise;
+		container.close();
+		const transitionContainer = await loadV2Container(firstSummaryAck.handle);
+		await waitForContainerConnection(transitionContainer);
+		await provider.ensureSynchronized();
+		const dObject = await requestFluidObject<DataObjectV1>(transitionContainer, "/");
+		dObject.containerRuntime.submitTransition();
+		const { summarizer, readonlyRuntime } = await deferred.promise;
+		const waitForSummarizerRuntimeConnection = new Deferred<void>();
+		summarizer.on("connected", () => {
+			waitForSummarizerRuntimeConnection.resolve();
+		});
+		await waitForSummarizerRuntimeConnection.promise;
```

Comment:
I think even in your PR we use this pattern of setting up listener to resolve differed multiple times. And it's somewhat wrong as you do not use once(), you use on() :) I suggest creating a general-purpose helper function that takes arbitrary EventEmitter, arbitrary event name and returns a promise when such event name is called. Implementation will create deferred and do what you do here, but for callers it would be a single line call - await waitEvent(emitter, "connected").